fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMerger.java

```diff
@@ -108,27 +112,70 @@ private CompletableFuture<Boolean> handleSuccess(final IndexDeferredMaintenanceC
         timeQuotaMillis = 0;
+        repartitionDocumentCount = 0;
         // Here: no errors, stop the iteration unless has more
-        final boolean hasMore = mergeControl.getMergesFound() > mergeControl.getMergesTried();
+        final boolean hasMore =
+                shouldGiveRepartitionSecondChance() ||
+                        mergeControl.getMergesFound() > mergeControl.getMergesTried();
         return  hasMore ? AsyncUtil.READY_TRUE : AsyncUtil.READY_FALSE;
     }
 
     private CompletableFuture<Boolean> handleFailure(final IndexDeferredMaintenanceControl mergeControl, Throwable e) {
         final FDBException ex = IndexingBase.findException(e, FDBException.class);
+        final IndexDeferredMaintenanceControl.LastStep lastStep = mergeControl.getLastStep();
         if (!IndexingBase.shouldLessenWork(ex)) {
             giveUpMerging(mergeControl, e);
         }
-        // Here: this exception might be resolved by reducing the number of merges or forcing shorter intervals between auto-commits
-        if (mergeControl.getMergesTried() < 2) {
-            handleSingleMergeFailure(mergeControl, e);
-        } else {
-            handleMultiMergeFailure(mergeControl, e);
+        switch (lastStep) {
+            case REBALANCE:
+                // Here: this exception might be resolved by reducing the number of documents to move during rebalancing
+                handlePotentialRebalanceFailure(mergeControl, e);
+                break;
+
+            case MERGE:
+                // Here: this exception might be resolved by reducing the number of merges or forcing shorter intervals between auto-commits
+                if (mergeControl.getMergesTried() < 2) {
+                    handleSingleMergeFailure(mergeControl, e);
+                } else {
+                    handleMultiMergeFailure(mergeControl, e);
+                }
+                break;
+
+            default:
+                giveUpMerging(mergeControl, e);
+                break; // "A switch statement does not contain a break", croaks a grumpy pmdMain.
         }
         return AsyncUtil.READY_TRUE; // and retry
     }
 
+    private void handlePotentialRebalanceFailure(final IndexDeferredMaintenanceControl mergeControl, Throwable e) {
+        repartitionDocumentCount = mergeControl.getRepartitionDocumentCount();
+        if (repartitionDocumentCount == -1) {
+            // Here: failed, despite being set to skip. No recovery.
+            giveUpMerging(mergeControl, e);
+        }
+        repartitionDocumentCount /= 2;
+        if (repartitionDocumentCount == 0) {
+            // Here: retry merge, but without partition re-balance
+            repartitionDocumentCount = -1;
+        }
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace(mergerLogMessage("IndexMerge: partition rebalance failure", mergeControl), e);
+        }
```

Comment:
I feel like this should be at least `DEBUG` level. 
```suggestion
        if (LOGGER.isDebugEnabled()) {
            LOGGER.debug(mergerLogMessage("IndexMerge: partition rebalance failure", mergeControl), e);
        }
```

Unless it's redundant, or would be logged many times during a deferred maintenance.